### DISCUSSION TITLE
proper way to use omp?

### DISCUSSION THREAD

**Discussion by retsyo** [pascalabcnet/pascalabcnet#2618] [Q&A]
the right output of the following code should be `664579`

however, none of the openmp's directive give `664579`. Any fix? Thanks

```pascal
const CACHE_SIZE: integer = 400;
var begin_count := 3;
var end_count := 10000000;
var cache_ptr := -1;
var primes_cache: array[0 .. CACHE_SIZE-1] of integer;

function IsPrimeNumber(n: integer): boolean;

  begin
  	var cur := 0;
//  	{$omp parallel for }
  	for var i := 0 to cache_ptr do
      begin
        	cur := primes_cache[i];
        	if (n mod cur) = 0 then 
            begin
              Result := false;
              exit;
            end;
      end;
  	
  	var m := floor(sqrt(n));
  	while(true) do
      begin
    		cur += 2;
    		if (cur > m) then break;
    		if (n mod cur ) = 0 then 
          begin
            Result := false;
            exit;
          end;
    	end;
  	
  	if (cache_ptr < CACHE_SIZE-1) then
      begin
    		cache_ptr += 1;
    		primes_cache[cache_ptr] := n;
      end;
  	
  	Result := true;
  end;


begin
  var count := 1;
  var tag: boolean;
  
  var t := Milliseconds;

//  {$omp parallel for }
  {$omp parallel for reduction(+:count)}
  for var num := begin_count to end_count do
    begin
      if num mod 2 = 0 then continue;
    	tag := IsPrimeNumber(num);
    	if (tag) then count += 1;
    end;
  
  printLn($'prime number count: {count}');
  
  Writeln($'total time: {Milliseconds - t} ms');
end.

```

**Comment by SunSerega** [comment]
1. Why did you use a static array for `primes_cache`? `primes_cache := new integer[CACHE_SIZE]` would be better in every way, since static arrays exist in this pascal only for compatibility and use regular dynamic arrays within.

2. `cache_ptr += 1` is not locked in any way. It is three operations of the processor (read; add 1; write), not even mentioning the processor's cache of RAM, which could be separate for each thread.

3. `$omp` is also outdated. It uses methods of `System.Threading.Tasks.Parallel`.
   
   1) `reduction` of `$omp` uses `lock`. So to write to `count ` one thread first acquires the lock, blocking every other thread that reached this line. Then changes `count`. Then allows one single other thread to do the same. What is `$omp` for, if you have a single thread-sized bottleneck?
      
      It's better to use `System.Threading.Interlocked` to change the value in multiple threads at the same time, without getting undetermined behavior.
      
4. It's better to first write a non-multithreaded algorithm with a cache like this, to make more solid-looking code. Then attempt again.
